USB: Dual Cursor Lightgun

RedDevilus opened the pull request:
### Description of Changes
<!-- Brief description or overview on what was changed in the PR -->
There seems to be an issue with how the MTGS handles the enumeration and drawing, updating for the replaced crosshairs icons that the users provides like in a .png format. Master would be random but tends to only show USB Device 1 crosshair. But now 2 at same time should work. Hopefully gets rid of https://github.com/PCSX2/pcsx2/issues/11423
![2025-05-23_03-36](https://github.com/user-attachments/assets/c42e1eeb-09fe-4a23-b5eb-14355de71483)
![2025-05-23_03-35](https://github.com/user-attachments/assets/311a20a5-4efe-456d-b026-7aa606bdf320)
![2025-05-23_04-28](https://github.com/user-attachments/assets/ec26bd44-1cd5-445b-a0f7-4effa1f1d5c3)

### Rationale behind Changes
<!-- Why were these changes made?  What problem does it solve / area does it improve? -->
The niche lightgun users also using niche dual replaced icons was certainly a bug as you get ghost inputs if not the texture was carefully drawn.
### Suggested Testing Steps
<!-- If applicable, including examples you've already tested with / recommendations for how to test further is very helpful! -->
Normal games work, but best in any game that uses GunCon2 aka lightgun games.
### Did you use AI to help find, test, or implement this issue or feature?
<!-- Answer yes or no. If you answer yes, please provide a brief explanation how. -->
No, this mess is my own.

RedDevilus committed:
USB: Dual Cursor Lightgun

There seems to be an issue with how the MTGS handles the enumeration and drawing, updating for the replaced crosshairs icons that the users provides like in a .png format. Master would be random but tends to only show USB Device 1 crosshair. But now 2 at same time should work. Hopefully gets rid of https://github.com/PCSX2/pcsx2/issues/11423

bslenul commented:
Thanks for the PR! ❤️ 

It seems to work fine for me with any renderer but GL, when using GL I'm getting a crash on game start:

[GL-crash.zip](https://github.com/user-attachments/files/20408099/GL-crash.zip)

lightningterror commented:
@bslenul Can you create a gs dump with a renderer that works where it would crash and post it here?

bslenul commented:
Sure, this is with "Automatic" (so VK in my case): [gun-gs-dump.zip](https://github.com/user-attachments/files/20409251/gun-gs-dump.zip)

The dump loads fine with all renderers expect GL (as long as I use at least 1 PNG as a crosshair).

lightningterror commented:
Dump works fine here, are you sure this dump crashes on gl?

![image](https://github.com/user-attachments/assets/d325a9fd-9acd-4524-8273-59390262d9f6)

bslenul commented:
It won't crash unless you've set a PNG crosshair, which I can't see on your screen.

Basically it looks like this for me with anything but GL:

![image](https://github.com/user-attachments/assets/bc8745dd-68f2-4cd7-a478-46f8ea857980)

lightningterror reviewed (changes requested):
UpdateSoftwareCursorTexture has to run on the GS thread for thread safety, otherwise it causes race conditions on OpenGL where you create texture without a valid context. All GL API calls must happen on the GS thread.

Try
```
	if (MTGS::IsOpen())
	{
		MTGS::RunOnGSThread([index]() {
			UpdateSoftwareCursorTexture(index);
		});
	}
```

bslenul commented:
Can confirm, no more crash with GL with that change! :+1:

bslenul commented:
Unrelated with that last change, but any idea why I'm getting this with this PR when running a Debug build?

![image](https://github.com/user-attachments/assets/176d3d33-7706-4d91-971d-3e6bd179ea0f)

![image](https://github.com/user-attachments/assets/70fad1e2-d866-4016-a2ff-4251ba34156f)

bslenul commented:
> Can confirm, no more crash with GL with that change! 👍

Argh, no more crash but now P2 cursor doesn't always appear again.

RedDevilus commented:
> > Can confirm, no more crash with GL with that change! 👍
> 
> Argh, no more crash but now P2 cursor doesn't always appear again.

Yup, if you have both crosshairs images with that fix in, it just reverts behaviour. Only shows first player crosshair and before showed both.

RedDevilus commented:
> Is this also related to #4472

This doesn't touch firewire and iLink, those are just stubbed aka placebo files in the codes still so we don't do that functionality.

It's just visual indicators for custom crosshairs which doesn't affect in-game parts.

RedDevilus committed:
USB: Fix MTGS thread undoing crosshairs

TheLastRar: '

This issue of the missing P2 cursor seems to be caused because the GS (and thus MTGS) gets reset shortly after USBopen, causing the command to get dropped.

Putting MTGS::WaitGS() at the end of this function (after the closing bracket for MTGS::RunOnGSThread()) seems enough to fix this.'

Didn't see where it lost the textures before but I did identify it was happening on the MTGS thread somehow. If you uncommented it, it did work but then you get issues with OpenGL renderer.

RedDevilus commented on pcsx2/ImGui/ImGuiManager.cpp:
```diff
@@ -1062,23 +1096,74 @@ void ImGuiManager::DrawSoftwareCursors()
+	// but may cause side effects like Player 2's crosshair overlaying everything in some games (e.g., Time Crisis 2/3). 
+	// Haven't really tested other games and also had issues even with showing inputs on my end in the OSD lower left section - RedDevilus
+	// This is a temporary workaround until a proper fix is implemented but it shouldn't break anything in-game (hopefully).
+	// MTGS::RunOnGSThread([index, image_path = std::move(image_path), image_scale, multiply_color]() {
```
That seems to have fixed the issue, double or single crosshairs now show up when needed and not that dance of half-brokeness. Thanks.

RedDevilus committed:
USB: Fix Big Picture locking out

FSUI waits on itself when switching/launching games in gamelist section but fine in Qt version.
